runtime/src/accounts_db.rs

```diff
@@ -2807,58 +2839,125 @@ impl AccountsDB {
+                let should_flush = !written_accounts.contains(account_key);
+                if should_flush {
+                    written_accounts.insert(*account_key);
+                } else {
+                    **account_bytes_saved += account.data.len();
+                    **num_accounts_saved += 1;
+                }
+                should_flush
+            } else {
+                // If we're not doing any cleaning, flush everything
+                true
+            }
+        };
 
-        // Flush all roots
-        let mut flush_roots_elapsed = Measure::start("flush_roots_elapsed");
-        let cached_roots = self.accounts_cache.clear_roots();
-        for root in &cached_roots {
-            self.flush_slot_cache(*root);
-            self.accounts_cache.set_max_flush_root(*root);
+        for root in cached_roots.iter().rev() {
+            if self.flush_slot_cache(*root, Some(&mut should_flush_f), true) {
+                num_roots_flushed += 1;
+            }
         }
 
         // Only add to the uncleaned roots set *after* we've flushed the previous roots,
         // so that clean will actually be able to clean the slots.
+        let num_new_roots = cached_roots.len();
         self.accounts_index.add_uncleaned_roots(cached_roots);
+        (num_new_roots, num_roots_flushed)
+    }
+
+    fn flush_accounts_cache(&self, force_flush: bool, max_clean_root: Option<Slot>) {
+        if !force_flush && self.accounts_cache.num_slots() <= MAX_CACHE_SLOTS {
+            return;
+        }
+
+        // Flush only the roots <= max_clean_root, so that snapshotting has all
+        // the relevant roots in storage.
+        let mut flush_roots_elapsed = Measure::start("flush_roots_elapsed");
+        let mut account_bytes_saved = 0;
+        let mut num_accounts_saved = 0;
+        let (total_new_cleaned_roots, num_cleaned_roots_flushed) = self.flush_rooted_slots_cache(
+            max_clean_root,
+            Some((&mut account_bytes_saved, &mut num_accounts_saved)),
+        );
         flush_roots_elapsed.stop();
 
         // Note we don't purge unrooted slots here because there may be ongoing scans/references
         // for those slot, let the Bank::drop() implementation do cleanup instead on dead
         // banks
 
         // If there are > MAX_CACHE_SLOTS, then flush the excess ones to storage
+        let (total_new_excess_roots, num_excess_roots_flushed) =
+            if self.accounts_cache.num_slots() > MAX_CACHE_SLOTS {
+                // Start by flushing the roots
+                //
+                // Cannot do any cleaning on roots past `max_clean_root` because future
+                // snapshots may need updates from those later slots, hence we pass `None`
+                // for `should_clean`.
+                self.flush_rooted_slots_cache(None, None)
+            } else {
+                (0, 0)
+            };
```

Comment:
how about putting this chunk at the end of `flush_rooted_slots_cache` (or my propose: `flush_rooted_accounts_cache`)? I think this code semantically belong to it. :)